pandas/core/internals/blocks.py

```diff
@@ -1692,15 +1695,34 @@ def _unstack(
         new_values = new_values.T[mask]
         new_placement = new_placement[mask]
 
-        blocks = [
-            # TODO: could cast to object depending on fill_value?
-            type(self)(
-                self.values.take(indices, allow_fill=allow_fill, fill_value=fill_value),
-                BlockPlacement(place),
-                ndim=2,
-            )
-            for indices, place in zip(new_values, new_placement)
-        ]
+        if allow_fill:
+            # needs_masking[i] calculated once in BlockManager.unstack tells
+            #  us if there are any -1s in the relevant indices.  When False,
+            #  that allows us to go through a faster path in 'take', among
+            #  other things avoiding e.g. Categorical._validate_scalar.
+            blocks = [
+                # TODO: could cast to object depending on fill_value?
+                type(self)(
+                    self.values.take(
+                        indices, allow_fill=needs_masking[i], fill_value=fill_value
+                    ),
+                    BlockPlacement(place),
+                    ndim=2,
+                )
+                for i, (indices, place) in enumerate(zip(new_values, new_placement))
+            ]
+        else:
```

Comment:
hmm can you avoid repeating this? (e.g. could create a `needs_masking[i]` with all values == False for the not allow_fill case

Reply:
sure